


#### Project Overview
<center>
Darmanis et al (2015) aimed to provide a map of the human brain transcriptome via the use of bioinformatics analysis. This project aimed to create a replicate analysis of this data, in order to determine whether the conclusions drawn by Darmanis et al (2015) could be replicated and whether the use of an alternative pipeline would alter any results. 

Our reanalysis of the Darmanis et (2015) data leads to the identification of a similar number of clusters, representing similar cell types. However, our initial reanalysis identified 9 key clusters, compared to 10 clusters from Darmanis et al (2015). After using an alternative pipeline including, the use of Seurat we identified 7 clusters. In both analyses, we did not identify the same gene clusters as identified by Darmanis et al (2015), indicating that the parameters and reference genome chosen were not the same as what was used in the initial analysis. Darmanis et al (2015) did fail to disclose some of the parameters used during their data analysis, indicating the reason why our work does differ. From this, it has been concluded, that although a successful replication and alternative analysis did occur, the resulting output is dependent on thee parameters chosen.
</center>


<center>


<img src="scRNA-overview.jpg" alt="scRNA Overview" width="400"/>

#### scRNAseq Overview
scRNAseq refers to sequencing on RNA at a single cell level. Implementing scRNAseq on transcriptomic data allows for advancements in our understanding of protein location, movement and interactions. As described in this figure, tissue is taken, then cells from this tissue are isolates and sequenced. Gene expression levels are then compared cell to cell, rather than via traditional methods of overall expression levels. This allows us to quantify where/what is expressing the protein and draw biological conclusions from that (Olsen and Baryawno, 2018). 
</center>

#### Members of the Group

**Contributors** :  Solomon, Megan, Ben , Karthik, Rawan 

